### DISCUSSION TITLE
JS cron job doesn't run

### DISCUSSION THREAD

**Discussion by denisgurev** [pocketbase/pocketbase#3252] [Q&A] [answered]
I'm using the JS hooks and tried to follow the docs but my cron job doesn't seem to run. I've already waited over an hour and it didn't print anything.

`cronAdd("sync", "1 * * * *", () => { console.log("run") })`

What I'm doing wrong?
I also can't find the cron jobs listed with `crontab -u dg -l`. Is this an OS issue or PocketBase one?

**Comment by ganigeorgiev** [comment, marked as answer]
I've tested it locally with PocketBase v0.17.7 and it works fine for me.

Are you sure that the above is the correct cron expression that you want?
Note that `1 * * * *` will fire **only on the first minute of every hour** - `15:01:00`, `16:01:00`, etc.
If you want your function to fire **on every minute** you can use `*` or "step" (`*/1`), aka. `* * * * *` - `15:01:00`, `15:02:00`, `15:03:00`, etc.

> I also can't find the cron jobs listed with crontab -u dg -l. Is this an OS issue or PocketBase one?

PocketBase doesn't use the OS level cron utility because it is not portable and there is no guarantee that it will work the same way everywhere. Instead we use a Go cron utility package based on a plain Go timer - https://github.com/pocketbase/pocketbase/tree/master/tools/cron.
Unless I've messed up something with the [cron expression parsing](https://github.com/pocketbase/pocketbase/blob/322508f6d197df9d0f7a7d380a8346518329259b/tools/cron/schedule.go#L138) (_see also [tests](https://github.com/pocketbase/pocketbase/blob/develop/tools/cron/schedule_test.go)_), I expect it to behave more-or-less the same as the OS `crontab` service as long as the application process is running.

If you are still not able to identify the issue:
- double check that your cron expression is correct (_if you are not sure, you can test it with [crontab.guru](https://crontab.guru/)_)
- ensure that your `pb_hooks` file ends in `.pb.js` (_or `pb.ts`; you can also add a `console.log` somewhere outside of the cron handler to make sure that the file is being loaded_);

**Reply by denisgurev** [reply]
Wow, I'm an idiot, sorry. Everything works fine - you are right, what I wanted was the "step" syntax, I saw `*/2` in the docs and thought that it was some typo or extra escape character :facepalm:.
As always thank you for the quick and detailed answer and sorry for wasting your time :heart:.

**Comment by shijinannuity** [comment]
Can we use cronAdd() inside a hook?

**Reply by ganigeorgiev** [reply]
No. Only as top-level function.

**Reply by ganigeorgiev** [reply]
@shijinannuity I'm sorry but it is not clear what is your use, nor what you are trying to do. Please don't treat the discussions as LLM chat.

If you want to register cron events dynamically one alternative is to create a "jobs collection" that will contains entry for each job that needs to be processed by a cron task, which will query that collection. But again without more details about your actual problem I don't think anyone will be able to assist you. If you have further questions please elaborate in a new Q&A discussion post.
